include/flatbuffers/grpc.h

```diff
@@ -23,50 +23,271 @@
+  }
+
+  inline ::grpc::Status VerifyGRPC() const {
+    if (Verify()) {
+      return ::grpc::Status::OK;
+    } else {
+      // DATA_LOSS: "Unrecoverable data loss or corruption."
+      return ::grpc::Status(::grpc::StatusCode::DATA_LOSS,
+                            "Message failed verification");
+    }
+  }
+
+  inline T *GetMutableRoot() {
+    return flatbuffers::GetMutableRoot<T>(mutable_data());
+  }
+
+  inline const T *GetRoot() const {
+    return flatbuffers::GetRoot<T>(data());
+  }
+
+  // This is only intended for serializer use, or if you know what you're doing
+  inline const grpc_slice &BorrowSlice() const { return slice_; }
+
+ private:
+  grpc_slice slice_;
+};
+
+class MessageBuilder;
+
+// SliceAllocator is a gRPC-specific allocator that uses the `grpc_slice`
+// refcounted slices to manage memory ownership. This makes it easy and
+// efficient to transfer buffers to gRPC.
+class SliceAllocator : public Allocator {
+ public:
+  inline SliceAllocator() : slice_(grpc_empty_slice()) {}
+
+  SliceAllocator(const SliceAllocator &other) = delete;
+  SliceAllocator &operator=(const SliceAllocator &other) = delete;
+
+  virtual inline ~SliceAllocator() { grpc_slice_unref(slice_); }
+
+  virtual inline uint8_t *allocate(size_t size) override {
+    assert(GRPC_SLICE_IS_EMPTY(slice_));
+    slice_ = grpc_slice_malloc(size);
+    return GRPC_SLICE_START_PTR(slice_);
+  }
+
+  virtual inline void deallocate(uint8_t *p, size_t size) override {
+    assert(p == GRPC_SLICE_START_PTR(slice_));
+    assert(size == GRPC_SLICE_LENGTH(slice_));
+    grpc_slice_unref(slice_);
+    slice_ = grpc_empty_slice();
+  }
+
+  virtual inline uint8_t *reallocate_downward(uint8_t *old_p, size_t old_size,
+                                              size_t new_size) override {
+    assert(old_p == GRPC_SLICE_START_PTR(slice_));
+    assert(old_size == GRPC_SLICE_LENGTH(slice_));
+    assert(new_size > old_size);
+    grpc_slice old_slice = slice_;
+    grpc_slice new_slice = grpc_slice_malloc(new_size);
+    uint8_t *new_p = GRPC_SLICE_START_PTR(new_slice);
+    memcpy(new_p + (new_size - old_size), old_p, old_size);
+    slice_ = new_slice;
+    grpc_slice_unref(old_slice);
+    return new_p;
+  }
+
+ private:
+  inline grpc_slice &get_slice(uint8_t *p, size_t size) {
+    assert(p == GRPC_SLICE_START_PTR(slice_));
+    assert(size == GRPC_SLICE_LENGTH(slice_));
+    return slice_;
+  }
+
+  grpc_slice slice_;
+
+  friend class MessageBuilder;
+};
+
+// SliceAllocatorMember is a hack to ensure that the MessageBuilder's
+// slice_allocator_ member is constructed before the FlatBufferBuilder, since
+// the allocator is used in the FlatBufferBuilder ctor.
+namespace detail {
+struct SliceAllocatorMember {
+  SliceAllocator slice_allocator_;
+};
+}
+
+// MessageBuilder is a gRPC-specific FlatBufferBuilder that uses SliceAllocator
+// to allocate gRPC buffers.
+class MessageBuilder : private detail::SliceAllocatorMember,
```

Comment:
I'd personally prefer to avoid the extra verbosity, because it will just be an extra line of boilerplate in every single RPC func. No one really sees this implementation detail, so maybe it's not so bad?